File: swift_vs_rust/match/enum.swift
//
// enum.swift
// Swift VS Rust
//
// Created by <Name> on <Date>.
// Copyright <Year and company>. All rights reserved.
//

import Swift

public enum ArithmeticOperation<ArithmeticValue: IntegerArithmeticType> {
  case Value(ArithmeticValue)
  indirect case Sum(ArithmeticOperation, ArithmeticOperation)
  indirect case Mul(ArithmeticOperation, ArithmeticOperation)

  public func solve() -> ArithmeticValue {
    switch self {
    case .Value(let value):      return value
    case .Sum(let lhs, let rhs): return lhs.solve() + rhs.solve()
    case .Mul(let lhs, let rhs): return lhs.solve() * rhs.solve()
    }
  }
}

public protocol Nameable {
  var name: String { get }
}

extension ArithmeticOperation: Nameable {
  public var name: String {
    switch self {
    case .Value(_):  return "Value"
    case .Sum(_, _): return "Sum"
    case .Mul(_, _): return "Mul"
    }
  }
}

let op = ArithmeticOperation.Sum(.Value(10), .Mul(.Value(20), .Value(2)))

print("Swift: op is \(op.name) solved \(op.solve())")

File: swift_vs_rust/match/enum.rs
enum MathOperation {
    Value(i32),
    Sum(Box<MathOperation>, Box<MathOperation>),
    Mul(Box<MathOperation>, Box<MathOperation>)
}

trait HasName {
    fn name(&self) -> &'static str;
}

impl HasName for MathOperation {
    fn name(&self) -> &'static str {
        match *self {
            MathOperation::Value(..) => "Value",
            MathOperation::Sum(..) => "Sum",
            MathOperation::Mul(..) => "Mul"
        }
    }
}

impl MathOperation {
    fn solve(&self) -> i32 {
        match *self {
            MathOperation::Value(i)         => i,
            MathOperation::Sum(ref left, ref right) => left.solve() + right.solve(),
            MathOperation::Mul(ref left, ref right) => left.solve() * right.solve()
        }
    }
}

fn main() {
    let op = MathOperation::Sum(Box::new(MathOperation::Value(10)),
                                Box::new(MathOperation::Mul(Box::new(MathOperation::Value(20)),
                                                            Box::new(MathOperation::Value(2)))));

    ;
    println!("Rust:  op is {} solved {}", op.name(), op.solve());
}
